Add error message when user tries to follow their own account

lenikadali opened the pull request:
Fixes #22690

Based on [this PR](https://github.com/mastodon/mastodon/pull/23373), we are adding a meaningful error message when a user tries to follow their own account.

lenikadali committed:
Add error message when user tries to follow their own account

Fixes #22690

Based on this PR https://github.com/mastodon/mastodon/pull/23373,
we are adding a meaningful error message when a user tries
to follow their own account.

lenikadali committed:
Add error message when user tries to follow their own account

Fixes #22690

Based on this PR https://github.com/mastodon/mastodon/pull/23373,
we are adding a meaningful error message when a user tries
to follow their own account.

mjankowski commented on app/controllers/api/v1/accounts_controller.rb:
```diff
@@ -47,6 +47,8 @@ def follow
     options = @account.locked? || current_user.account.silenced? ? {} : { following_map: { @account.id => { reblogs: follow.show_reblogs?, notify: follow.notify?, languages: follow.languages } }, requested_map: { @account.id => false } }
 
     render json: @account, serializer: REST::RelationshipSerializer, relationships: relationships(**options)
+  rescue FollowService::SelfFollowError
```
Is this ever actually raised?

lenikadali replied:
Good catch. Let me go over it and make sure that it is raised 👍🏿

mjankowski replied:
One idea here -- could we re-use `Mastodon::ValidatorError` (with a custom string)?

I think this could keep the entire check at the controller level, would match existing style, and would let you remove all the new error classes, along with the changes to the service and service spec.

For an example of what I mean - look at `api/v1/filters#update` and how it raises that class (or `check_accounts_limit` in this controller, though it does not have a custom message).

I suspect you could keep your request spec changes here, but move the actual check to a `before_action` in the controller (since it knows what accounts are relevant before passing in).

mjankowski commented on app/services/follow_service.rb:
```diff
@@ -5,6 +5,9 @@ class FollowService < BaseService
   include Payloadable
   include DomainControlHelper
 
+  class Error < StandardError; end
```
Can we re-use some existing classes here with custom messages or i18n strings?

lenikadali replied:
Sure! Will look through the codebase and see if I can identify an appropriate class to use.

lenikadali replied:
See comment above regarding i18n strings.

mjankowski commented on spec/requests/api/v1/accounts_spec.rb:
```diff
@@ -147,6 +147,27 @@
+      end
+
+      it 'returns JSON with an error message' do
+        json = body_as_json
```
You might want to rebase this -- the `body_as_json` helper is no longer available (use `response.parsed_body` directly).

Separatetly, you can you combine/chain the three assertions together so we can assert multiple things about the same response (avoid multiple round trips, multiple factories, etc).

lenikadali replied:
Noted. Will do.

mjankowski commented on spec/services/follow_service_spec.rb:
```diff
@@ -136,6 +136,16 @@
+          .to raise_error(FollowService::SelfFollowError)
+          .and(not_change { sender.following?(sender) })
+
+        expect(sender.following?(sender)).to be false
```
This could be an `.from(false)` added to the end of the `not_change` above?

lenikadali replied:
Hmmm... I can try it out and see 🙂

lenikadali committed:
Add review changes

Added the following review changes:

raise the error using i18n string
Use response.parsed_body directly and chain assertions

lenikadali committed:
Add review changes

Added the following review changes:

raise the error using i18n string
Use response.parsed_body directly and chain assertions

lenikadali commented on app/controllers/api/v1/accounts_controller.rb:
```diff
@@ -47,6 +47,8 @@ def follow
     render json: @account, serializer: REST::RelationshipSerializer, relationships: relationships(**options)
+  rescue FollowService::SelfFollowError
+    render json: { error: I18n.t('Following your own account is not allowed') }, status: 403
   end
```
I've opted to use the i18n string here since that fits the pattern used in the codebase. Let me know if there's a specific error class I should have used.

lenikadali reviewed (commented):
Hi @mjankowski 

I think I've addressed all the review comments so this is good for another review.

Edit: Not sure if I should rebase or merge in order to update this branch. Please let me know which style Mastodon uses 🙏🏿

mjankowski commented:
> I think I've addressed all the review comments so this is good for another review.

Will look, thanks.

> Edit: Not sure if I should rebase or merge in order to update this branch. Please let me know which style Mastodon uses 

Accepted PRs wind up being squashed down into one commit with the PR title as the commit title ... so basically, whatever is easiest for you to keep the branch updated is fine.

mjankowski commented on app/controllers/api/v1/accounts_controller.rb:
```diff
@@ -47,6 +47,8 @@ def follow
 
     render json: @account, serializer: REST::RelationshipSerializer, relationships: relationships(**options)
+  rescue FollowService::SelfFollowError
+    render json: { error: I18n.t('Following your own account is not allowed') }, status: 403
```
Using `I18n` here is good ... but it looks like this is just putting the error string right here, instead of adding an i18n key?

lenikadali replied:
Hmmm... okay so I'm not sure what adding an i18n key means. I had assumed that doing the above is what was meant. Are there any examples of using an `i18n` key that I can look at in the codebase?

ClearlyClaire replied:
Hi! I18n strings are defined in `config/locales`. One example you can have a look at is `invites.invalid`, defined in `config/locales/en.yml` and used in `app/controllers/api/v1/invites_controller.rb`

lenikadali committed:
Merge branch 'main' into bug/add-error-message-when-user-tries-to-follow-themselves

lenikadali committed:
Merge remote-tracking branch 'leni/bug/add-error-message-when-user-tries-to-follow-themselves' into bug/add-error-message-when-user-tries-to-follow-themselves

lenikadali committed:
Change to use a before_action, add i18n string

Changed the implementation to use a before_action within
the controller and (as a result) remove the need for using
the FollowService.

lenikadali committed:
Update message to be the one we expect in test

lenikadali committed:
RuboCop fixes

lenikadali committed:
Add subject to test

Added subject to test because it seems that without
a re-declaration of subject, the response will be nil

As of now, the response is 422. Not sure why.

lenikadali committed:
Forgot to remove unneeded Error declaration

lenikadali commented:
Hey @mjankowski 👋🏿 

I've made the changes as discussed [here](https://github.com/mastodon/mastodon/pull/31910#discussion_r1767528339) and [here](https://github.com/mastodon/mastodon/pull/31910#discussion_r1767536660)

The current issue is that [the test](https://github.com/mastodon/mastodon/blob/d36c2f27f3841b230d65b13cac998748ad84b5c8/spec/requests/api/v1/accounts_spec.rb) currently fails as it returns a 422 code and no response body.

Unfortunately, I am more used to using Minitest and `binding.pry` so not sure how to debug how/why the test is failing with this error.

mjankowski commented on spec/requests/api/v1/accounts_spec.rb:
```diff
@@ -163,6 +163,31 @@
+      it 'returns http forbidden and error message' do
+        subject
+
+        error_msg = 'Following your own account is not allowed'
```
This can be pulled from I18n as well, instead of repeated (ie, remove this local variable and just use the `I18n.t(...)` block in the assertion below.

lenikadali replied:
Good catch. Have modified the test accordingly 👍🏿

mjankowski commented on spec/requests/api/v1/accounts_spec.rb:
```diff
@@ -163,6 +163,31 @@
+        expect(response.parsed_body)
+          .to include(
+            error: error_msg,
+            following: false
```
Is this param (following) actually in the json?

lenikadali replied:
No it isn't. I think it was based on the assumption that there are some additional JSON fields that we return in addition to the error message. Have removed it.

mjankowski commented on app/services/follow_service.rb:
```diff
@@ -50,7 +50,7 @@ def mark_home_feed_as_partial!
   end
 
   def following_not_possible?
-    @target_account.nil? || @target_account.id == @source_account.id || @target_account.unavailable?
```
The `FollowService` could be (and is) called from other places as well, not just the controller being changed here - should probably leave this in?

lenikadali replied:
Have restored the method to what it was originally. If there's something I've missed, the specs should catch it 😅

lenikadali committed:
Merge branch 'main' into bug/add-error-message-when-user-tries-to-follow-themselves

lenikadali committed:
Restore following_not_possible, directly render error

Restored the following_not_possible method to its original
version since it is (likely) called in other places which
rely on its original behaviour

Switched to directly rendering the error message if we find
that someone is trying to follow their account.

Updated the error expectation in the spec to match the response
the before_action returns

lenikadali commented on app/controllers/api/v1/accounts_controller.rb:
```diff
@@ -47,6 +48,8 @@ def follow
     options = @account.locked? || current_user.account.silenced? ? {} : { following_map: { @account.id => { reblogs: follow.show_reblogs?, notify: follow.notify?, languages: follow.languages } }, requested_map: { @account.id => false } }
 
     render json: @account, serializer: REST::RelationshipSerializer, relationships: relationships(**options)
+  rescue Mastodon::ValidationError
```
Doing this worked. Thank you! 🙌🏿

lenikadali committed:
RuboCop fix for Style/IfUnlessModifier

lenikadali commented:
@mjankowski I think this is ready for merge.
Let me know if I missed something 😅 or there is something I can improve in terms of implementation. Thank you 🙏🏿

ClearlyClaire reviewed (dismissed):
Thanks for your contribution! This looks good to me!

ClearlyClaire commented on config/locales/en.yml:
```diff
@@ -11,6 +11,7 @@ en:
       one: Follower
       other: Followers
     following: Following
+    self_follow_error: Following your own account is not allowed
```
Oh, I missed this was not inserted at the correct place. You can fix that by running `bundle exec i18n-tasks normalize`

lenikadali committed:
Normalize locale file

oneiros commented:
@lenikadali thank you very much for your contribution (and your patience)!
